source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -47,9 +64,38 @@ void LocalRateLimiterImpl::onFillTimer() {
   } while (
       !tokens_.compare_exchange_weak(expected_tokens, new_tokens_value, std::memory_order_relaxed));
 
+  onFillTimerDescriptorHelper();
   fill_timer_->enableTimer(fill_interval_);
 }
 
+void LocalRateLimiterImpl::onFillTimerDescriptorHelper() {
+  auto cur = t_logclock::now().time_since_epoch().count();
+  for (const auto& descriptor : last_refill_per_descriptor_) {
+    if ((cur - descriptor.second) >
+        convertRateLimitUnitToMilliseconds(descriptor.first.limit_.value().unit_)) {
+      auto it = tokens_per_descriptor_.find(descriptor.first);
+      if (it == tokens_per_descriptor_.end()) {
+        continue;
+      }
+
+      uint32_t expected_tokens = it->second->tokens.load(std::memory_order_relaxed);
+      uint32_t new_tokens_value;
+      do {
+        // expected_tokens is either initialized above or reloaded during the CAS failure below.
+        new_tokens_value = descriptor.first.limit_.value().requests_per_unit_;
+
+        // Testing hook.
+        synchronizer_.syncPoint("on_fill_timer_descriptor_pre_cas");
```

Comment:
Very quick drive by pre-full code review: This code is complicated and should not be duplicated here and elsewhere. Refactor to have a single copy of the complicated logic.

Reply:
Sure..